src/6model/reprs.c

```diff
@@ -271,58 +275,54 @@ void MVM_repr_initialize_registry(MVMThreadContext *tc) {
     register_core_repr(CStr);
     register_core_repr(CArray);
     register_core_repr(CStruct);
-    register_core_repr(CUnion);
     register_core_repr(ReentrantMutex);
     register_core_repr(ConditionVariable);
     register_core_repr(Semaphore);
     register_core_repr(ConcBlockingQueue);
     register_core_repr(AsyncTask);
     register_core_repr(Null);
-    register_core_repr(CPPStruct);
     register_core_repr(NativeRef);
+    register_core_repr(CUnion);
     register_core_repr(MultiDimArray);
+    register_core_repr(CPPStruct);
     register_core_repr(Decoder);
-    register_core_repr(SpeshLog);
     register_core_repr(StaticFrameSpesh);
     register_core_repr(SpeshPluginState);
 
-    tc->instance->num_reprs = MVM_REPR_CORE_COUNT;
+    assert(tc->instance->num_reprs == MVM_REPR_CORE_COUNT);
 }
 
-static MVMReprRegistry * find_repr_by_name(MVMThreadContext *tc,
-        MVMString *name) {
-    MVMReprRegistry *entry;
+/* Get a representation's ID from its name. Note that the IDs may change so
+ * it's best not to store references to them in e.g. the bytecode stream. */
+MVMuint32 MVM_repr_name_to_id(MVMThreadContext *tc, MVMString *name) {
+    if (!MVM_str_hash_key_is_valid(tc, name)) {
+        MVM_str_hash_key_throw_invalid(tc, name);
+    }
 
     uv_mutex_lock(&tc->instance->mutex_repr_registry);
-    MVM_HASH_GET(tc, tc->instance->repr_hash, name, entry)
-
-    if (entry == NULL) {
+    MVMuint32 idx = MVM_index_hash_fetch_nt(tc, &tc->instance->repr_hash, tc->instance->repr_names, name);
+    if (idx == MVM_INDEX_HASH_NOT_FOUND) {
         char *c_name = MVM_string_ascii_encode_any(tc, name);
         char *waste[] = { c_name, NULL };
         uv_mutex_unlock(&tc->instance->mutex_repr_registry);
-        MVM_exception_throw_adhoc_free(tc, waste, "Lookup by name of unknown REPR: %s",
-            c_name);
+        MVM_exception_throw_adhoc_free(tc, waste, "Lookup by name %p of unknown REPR: %s",
+                                       name, c_name);
```

Comment:
Not quite sure printing the pointer address of the `MVMString` adds anything to this error output?

Reply:
Agree. That should go.

Reply:
Agree. That should be neuralised.